### PR TITLE
Bootstrap Django API

### LINKED ISSUE #414: Add Django to development environment
- Add a Django application, backed by Postgresql, to the project. We haven't set up a Django / React project from scratch in a while, so it would be a good ideal to consult with someone on CA and then copy from their best-practice project as much as possible.
- `scripts/setup`, `scripts/update`, and `scripts/server` should all work as expected (i.e. they should get the project working and run migrations as appropriate).
- Also add a `scripts/manage` to issue Django management commands.
- Choose a port for Django runserver and add it to the team ports document.
- You don't need to take any steps to implement CI or infrastructure.
- The frontend should be able to hit a basic API endpoint implemented in Django and confirm success. This doesn't have to do anything useful except confirm that the React frontend has successfully communicated with the Django API. I expect that we'll want the React app and the Django app to function from the same domain (rather than communicating from localhost:9966 for the React app to localhost:8000 for Django, for example) but consult with CA on this.

### PR DESCRIPTION
**Pull request by aaronxsu**
## Overview

This PR:
- adds `django` and `database` services for local dev as API and DB
- adds/updates `./scripts/bootstrap`, `./scripts/setup`, `./scripts/update`, `./scripts/manage`, `./scripts/server`, `./scripts/dbshell`
- adds a simple `/api/` API route and returns a message that shows the API works as expected
- adds a small test on the frontend to allow testing the communication with local dev API

### Notes

1. I made some notes as single comments in the files changed below. Feel free to discuss!
2. I did not add in Vagrant in this solution, and would like to propose to start from just using Docker. If we find it easier to add Vagrant as a layer in between in the future, I'd be happy to take that issue as well.

## Testing Instructions

 * Make sure your local docker app is running
 * At the root of the directory, run `./scripts/setup`, which should bootstrap the app, install deps, build containers, and run migrations
 * Run `./scripts/server`, which should spin up a local DB, frontend, and Django API server
 * After the Django API is up and running, make sure the following is successful
```
curl --location --request GET 'http://localhost:8081/api'
```
* Go to http://localhost:9966, open your network tabs in the dev tools, and filter to the `/api/` call. Make sure the request is also successful
* At the root of the directory, run `./scripts/manage migrate` and make sure it runs migrations, although there is no migration to run now, which is fine
~At the root of the directory, run `./scripts/dbshell` and make sure it takes you to a local `psql` shell and you could do normal SQL stuffs there~

Resolves #414

### CODE REVIEW

File: backend/Dockerfile
Comment (aaronxsu) [inline backend/Dockerfile]: Let's use `django 3.2` and `python 3.9` for this project API!

File: backend/Dockerfile
Comment (aaronxsu) [inline backend/Dockerfile]: client for `postgres 13`

File: backend/echo/settings.py
Comment (aaronxsu) [inline backend/echo/settings.py]: Let's use the `postgres-postgis` engine

File: backend/api/views.py
Comment (aaronxsu) [inline backend/api/views.py]: This is just for testing purpose for now to make sure that the API is working as expected. We should remove this in the future.

File: scripts/bootstrap
Comment (aaronxsu) [inline scripts/bootstrap:24]: I am assuming that we would pull some env vars from S3 here. Please let me know if this should be something else.

File: scripts/dbshell
Comment (aaronxsu) [inline scripts/dbshell]: The hope here is to let us get into a local db shell and see what is up in there if we'd like to check something quickly.

File: README.md
Comment (aaronxsu) [inline README.md:83]: Please let me know if these suffice. I am happy to add more scripts or more descriptions of other scripts that were written before.

File: backend/echo/settings.py
Comment (aaronxsu) [inline backend/echo/settings.py]: Though it looks like OAR did something fancy here: https://github.com/open-apparel-registry/open-apparel-registry/blob/develop/src/django/oar/settings.py#L380-L399 -- If this whitelist won't suffice, I am happy to emulate what OAR does in the above linked code block.

File: backend/requirements.txt
Comment (aaronxsu) [inline backend/requirements.txt]: I think these might be the bare minimum.

File: docker-compose.yml
Comment (aaronxsu) [inline docker-compose.yml]: I saw the requirement to "add to team ports document"- could you let me know where you find that doc?

File: docker-compose.yml
Comment (aaronxsu) [inline docker-compose.yml:23]: I think we would need an `.env` file at a point for these env vars.

File: README.md
Comment (ddohler) [inline README.md]:
```suggestion
| `manage`   | Running Django manage.py commands    |
```

As noted below, we don't really need this as a separate command if we already have `./scripts/manage` available, so I think we can just document `manage` and call it good.

File: scripts/server
Comment (ddohler) [inline scripts/server:23]: Nice catch! I tried to figure out why this was put here in the first place, but couldn't. Taking it out doesn't seem to hurt anything that I can find, and it certainly speeds up development because the frontend container doesn't have to get rebuilt every time you start development. If we do find that removing this causes problems, hopefully we can find a different solution than just reverting this change, because development is a lot more pleasant now that `./scripts/setup` doesn't take several minutes to get going.

File: scripts/update
Comment (ddohler) [inline scripts/update]: I think we need to add `docker-compose stop` at the end here so that this doesn't leave the containers running after the script finishes.

File: scripts/update
Comment (ddohler) [inline scripts/update]: Since we have `./scripts/manage` already available, another way to handle this would be to only bring up the `database` service on line 25 and then use `./scripts/manage migrate` here.

File: scripts/update
Comment (ddohler) [inline scripts/update:31]:
```suggestion
        docker-compose up -d database
```
(To match the comment below)

File: scripts/update
Comment (ddohler) [inline scripts/update:40]: I think static files will need to be accounted for here as well; they're necessary for the Django admin to function correctly and for the Django Rest Framework browsable API. I think we'll need to add `./scripts/manage collectstatic --no-input` here and [adjust `settings.py`](https://docs.djangoproject.com/en/3.2/howto/static-files/).

File: scripts/bootstrap
Comment (ddohler) [reply scripts/bootstrap:24]: Hmm, this is an interesting point. The backend of the application doesn't need any environment variables yet, and the frontend environment variables are managed by this section in the README:
> To start developing, create a set of Taui environment variables for development:
>
> ```
> $ cp taui/configurations/default/env.yml.tmp taui/configurations/default/env.yml
> ```
>
> Make sure to edit `env.yml` to set the appropriate secrets for development.

So I think probably the "correct" thing to do here (in quotes because I'm not sure it's actually the best thing to do) would be to have this script perform those instructions, i.e., copy the `env.yml` file and fill in the appropriate values. However, that seems like it's kind of outside of the scope of this task because the `env.yml` file is only related to frontend concerns; it won't impact Django, and the Django app currently doesn't need a separate `.env` file for local development anyway.

Additionally, if we ever do need an `.env` file for local development, we'll then have two different types of environment files for the frontend and backend (`env.yml` and `.env`, respectively). That seems unnecessarily complicated, and potentially confusing or harmful, so I think we would want to prioritize eliminating the difference, which we could do by [migrating away from `mastarm`](https://github.com/azavea/echo-locator/issues/376).

So to summarize, I think we can probably skip adding this file for now (or leave it as a placeholder) because the most useful thing it could do at the moment (copying env.yml) isn't related to Django and might not last very long anyway. We should probably update #376 with a note about this discussion since this seems like a point in favor of doing that.

How does that sound?

File: docker-compose.yml
Comment (ddohler) [reply docker-compose.yml]: Sent a link to the ports document out of band; this is just a reminder to update this file to match whatever port we choose.

File: docker-compose.yml
Comment (ddohler) [inline docker-compose.yml:56]: I'm not sure that this is having the intended effect. When I bring up the project from scratch (which can be simulated by running `docker-compose stop && docker-compose down`), I consistently receive an error connecting to the database:

```
django.db.utils.OperationalError: connection to server at "database" (172.25.0.2), port 5432 failed: Connection refused
	Is the server running on that host and accepting TCP/IP connections?
```

Everything works fine on the second try, however, so I'm not sure whether this is a race condition or some kind of different behavior that occurs when creating the container for the first time.

File: scripts/dbshell
Comment (ddohler) [reply scripts/dbshell]: Django makes `manage.py dbshell` available, so we can already get into the database via `./scripts/manage dbshell`. Having a separate script does save some typing, but I'm not sure the extra overhead of maintaining another script is worth it to save a few characters. I'd probably just remove this file, but I'm not opposed to keeping it if you think it would be helpful.

File: backend/api/views.py
Comment (ddohler) [inline backend/api/views.py]: Looks like this is a leftover from somewhere else?

File: backend/api/views.py
Comment (ddohler) [reply backend/api/views.py]: This would be great as a comment in the code as well, so that future visitors know that this is safe to remove.

File: backend/echo/settings.py
Comment (ddohler) [reply backend/echo/settings.py]: I think the differences with OAR extend beyond just the different CORS configuration. The CORS configuration on OAR is actually not relevant for the main OAR site, because OAR serves the static frontend and the API from the same host and port (e.g. while visiting `https://openapparel.org/` one might see a request to `https://openapparel.org/api/contributor-lists/?contributors=2422`). This means that CORS doesn't apply at all for the application's frontend because the API requests aren't cross-origin in the first place. I believe the fancier setup in OAR is to allow use of the OAR API from within certain other websites that have some kind of relationship with OAR, which isn't really something we expect to do.

The current setup here, by contrast, is cross-origin because the API and frontend are hosted on different ports, and if we continue that architecture through to deployment, we'll end up with the API hosted on something like `api.echosearch.org/api/` while the rest of the app is hosted at `www.echosearch.org`, and then we'll have to add `*.echosearch.org` to this whitelist.

I see a few different ways to proceed from here.

If we want to get closer to how OAR does things, OAR uses `django-spa` and `whitenoise` to serve the static files from within the Django app, and they use Webpack's proxy feature to pass through requests to `/api/` and `/admin/` through to the Django server. Mastarm supposedly [has a proxy feature](https://github.com/azavea/mastarm#build), but it's not clear how well it works, so it may be that to use proxying in the build tool we'd need to [migrate away from Mastarm](https://github.com/azavea/echo-locator/issues/376).

Alternatively, we could stick with the status quo and commit to making the app cross-origin for the foreseeable future. This would avoid the work of upgrading Mastarm, and would allow us to leave the existing deployment scheme for static files largely untouched. If we switched to hosting the files in Django, we'd need to swap out the existing S3-based hosting with Django static hosting for everything. However, this would potentially make the app more confusing for administrative users because the Django admin would have to be accessed at something like `api.echosearch.org/admin/`; if we put everything on the same domain it could be `www.echosearch.org/admin/` which seems friendlier.

Overall, Mastarm has been a continuing source of confusion and frustration (including one other thread in this PR), so I'm leaning toward the same-origin solution so that we have another excuse to get rid of it, but I could also see sticking with a cross-origin solution to avoid the need to rework the static deployment (even in an imaginary world where we were already using a more standard bundler).

So I think my suggested path forward would be to check out Mastarm's proxy feature and try to add `django-spa` if it seems to work. If that all works out, then we can eliminate `CORS` support entirely. If we run into roadblocks then I think we can leave this as-is and prioritize removing Mastarm. How does that sound to you?

**Review by ddohler — CHANGES_REQUESTED** [review]
Thanks for putting this together! Overall, this looks really good. I found some smaller things that I think need to be addressed, and some larger architecture discussion items that might potentially lead to more work, or to other issues being created. Let me know what you think!

File: scripts/dbshell
Comment (aaronxsu) [reply scripts/dbshell]: Agreed- makes sense to me. I will remove this script.

File: scripts/server
Comment (aaronxsu) [reply scripts/server:23]: Thanks. This sounds good to me!

File: README.md
Comment (aaronxsu) [reply README.md]: Sounds good. I will remove it 👍🏼

File: backend/api/views.py
Comment (aaronxsu) [reply backend/api/views.py]: Oops. Yes, I will remove it.

File: backend/api/views.py
Comment (aaronxsu) [reply backend/api/views.py]: Makes sense. Lemme add a todo note in here so that we will remove it later.

File: scripts/bootstrap
Comment (aaronxsu) [reply scripts/bootstrap:24]: Ah, this is a very good point. I agree- it does not make much sense to me either to copy the `env.yml`from the template file and populate the values, since:1. we would potentially move away from `mastarm`; 2. the values would need some human op to grab from last pass and fill in manually. And yea, we indeed do not need any env vars for the Django dev yet. I am with you on leaving this `bootstrap` script as is and treat the current code comment as a placeholder!

File: scripts/update
Comment (aaronxsu) [reply scripts/update]: Smart choice and less verbose. I like it! I will update this.

File: scripts/update
Comment (aaronxsu) [reply scripts/update]: Ah, good point. I like this, since this will be cleaner. Will make this change!

File: scripts/update
Comment (aaronxsu) [reply scripts/update:40]: Good catch. I will add this one as well.

File: docker-compose.yml
Comment (aaronxsu) [reply docker-compose.yml:56]: Hmmm, that looks odd indeed. I tried running `docker-compose stop && docker-compose down` and then `./scripts/setup` and then `./scripts/server`. However, I did not find the above error message in the logs. Did I miss anything step here? 🤔

File: docker-compose.yml
Comment (aaronxsu) [reply docker-compose.yml]: Sounds good. I ended up using port `8085` for the Django server. Here is the PR for updating the docs: https://github.com/azavea/geospatial-apps/pull/261

File: backend/echo/settings.py
Comment (aaronxsu) [reply backend/echo/settings.py]: Yes, this plan sounds very good to me. I think the reasoning here makes a lot of sense. 

I do think we should start by testing if adding `django-spa` and `whitenoise` to serve static files and also update the `proxy` flag for `mastarm build` works. If this won't work as expected, I am with you on the fact that switching away to another bundler is a wise choice here since `mastarm` is not maintained and we now have a chance to make the switch since we need to make the local frontend dev work with a Django API. I do not think that the currently implemented `CORS` solution is a way to go considering the downstream build and infra overhead.

I did take a stab at adding `django-spa` and `whitenoise`, plus modifying the `--proxy` flag in [this commit](https://github.com/azavea/echo-locator/pull/419/commits/432b9113b09b55deb83ccd8d60d0759d4a228ea4), but I think I was missing something there and it did not work for me locally. I think it might be that it expects to find a `index.html` in `backend/static`, but `mastarm build` does not seem to produce such a thing while only produces something like `taui/assets/index.js`. Plus, I think we will need a way to mount the directory of bundled frontend result to where the django docker container can find in the `backend/static`, but I could be wrong.

Thus, given the above mentioned confusion, I am happy to pair with you when you get a chance to have this figured out together!

Comment (aaronxsu) [reply backend/echo/settings.py]: Update: discussed about the above on Slack, and it [this commit](https://github.com/azavea/echo-locator/pull/419/commits/c09e13776a1ea54109463fc12f4d7d19d14e175d) fixed the issue for local development 🎉

File: README.md
Comment (ddohler) [inline README.md:77]: We should also include docs for `manage` here.

File: backend/api/views.py
Comment (ddohler) [inline backend/api/views.py]: Our normal team practice is that all TODOs should have a link to a GitHub issue so that they aren't forgotten. We didn't have an issue for this, so I went and [made one](https://github.com/azavea/echo-locator/issues/422), but the more I'm thinking about it, the more I think the best thing to do is just to remove this before merging, because it's more of a debugging check to prove that everything works than something that needs to live beyond this PR.

File: taui/src/index.js
Comment (ddohler) [inline taui/src/index.js]: Same comment here as above, this is probably best removed before merging, that way we can be certain not to lose track of it later.

File: scripts/bootstrap
Comment (ddohler) [inline scripts/bootstrap:2]: This could use a TODO comment with a link to #376 since I think that is where this would be most likely to get fleshed out.

File: backend/Dockerfile
Comment (ddohler) [inline backend/Dockerfile]: One thing I forgot to mention in the previous review (sorry!) that I think would be good to do is make the directory structure match our standard structure a bit more closely. Normally we nest application code under `src/`, but that didn't matter so much when there was only a frontend to this app. However, now that we've got both a backend and a frontend, I think it would be best to try to work our way back to our standard structure. Moving the frontend might break some things, so if we can just start by putting the Django API inside of `src/` now, then we can move `taui/` into the `src/` folder as part of #376 . How does that sound?

File: docker-compose.yml
Comment (ddohler) [reply docker-compose.yml:56]: Okay, I think I see what's going on: the `healthcheck` option [is valid for v2.1 Compose files, but was _removed_ for v3 Compose files](https://github.com/peter-evans/docker-compose-healthcheck/issues/3) (which we're using). So I don't think that the Django container is waiting for the database container at all. I'm not sure why it's working for you and not for me, but I suspect it's just a difference between our two systems such that the database container comes up a bit slower on my system and exposes the race condition.

It looks like the easiest solution may be to simply switch the compose file syntax version to `2.1` instead of `3`. That seemed to solve the issue for me.

File: scripts/update
Comment (ddohler) [inline scripts/update:39]: The database connection errors that I was getting exposed that this doesn't stop the database container when there's an error. I was able to get the containers to clean up correctly by adding this to the script:

```
function cleanup() {
  docker-compose stop
}

trap cleanup ERR
```

File: taui/src/index.js
Comment (ddohler) [inline taui/src/index.js]: No need to change this if you're going to remove this whole thing, but this can be just `/api/` since this app isn't operating cross-origin any longer:
```suggestion
    fetch('/api/').then(function (response) {
```

File: taui/package.json
Comment (ddohler) [inline taui/package.json:17]: One thing I noticed is that if you try to access `/api/` through the proxy (i.e. `localhost:9966/api/`) the static files don't show up and return 404s. However, I don't think we need to fix that because we're planning to shift away from Mastarm anyway, and everything works fine if you access the Django container directly via `localhost:8085/api/`.

**Review by ddohler — APPROVED** [review]
This is working well! I found a few more things that I think should be changed but it looks like they've all got pretty easy resolutions, so this is good to go after cleaning those up!

File: backend/api/views.py
Comment (aaronxsu) [reply backend/api/views.py]: Thanks for creating the issue. It makes sense to me to remove this API endpoint at this time.

File: taui/src/index.js
Comment (aaronxsu) [reply taui/src/index.js]: Sounds good!

File: backend/Dockerfile
Comment (aaronxsu) [reply backend/Dockerfile]: That sounds like gold to me! I am happy to follow our usual practice here.

**Pull request merged**
